### ISSUE TITLE
introduce HKT to vavr

### ISSUE THREAD

**Issue by hepin1989** [vavr-io/vavr#2290]
support HKT encoding in Java,helpful to building Effect like thing in Java

**Comment by CauchyPeano** [comment]
How do you think it will be possible to do this in current version of Java?

**Comment by danieldietrich** [comment]
Thanks for the suggestion. In Javaslang there were Kind1 and Kind2 for that purpose. But they had several caveats and I removed them. I was invited as an admin for 
https://github.com/KindedJ/KindedJ, a library that wants to provide standard HKT interfaces for Java but I decided not to pollute Vavr’s API with this kind of abstractions.

Simply spoken, it isn’t currently possible in Java to implement higher-kinded types that work across deep type hierarchies. As a responsible library designer, I can’t add such a non-working feature into Vavr’s code base.

Another reason is my strategic decision to keep algebraic / math related API out of Vavr. Our functionality is inherently algebraic. I don’t see the necessity to move to a non-existing audience that does purely functional programming in Java using algebraic abstractions.
